[enh] add settings option to enable/disable search formats

return42 opened the pull request:
## What does this PR do?

- eff337a4 [enh] add settings option to enable/disable search formats

## How to test local

In settings.yml edit

    search:
        ...
        formats: [html]

The denied formats should no longer available in the link box:

![image](https://user-images.githubusercontent.com/554536/119707876-2eaf5800-be4b-11eb-84e8-b989c055c36b.png)

Request a format: ( http://127.0.0.1:8888/search?q=foo&format=csv ) should result in a 403

![image](https://user-images.githubusercontent.com/554536/119708269-85b52d00-be4b-11eb-812f-d061889033a8.png)

## Related issues

Closes #95

return42 committed:
[enh] add settings option to enable/disable search formats

Access to formats can be denied by settings configuration::

    search:
        formats: [html, csv, json, rss]

Closes: https://github.com/searxng/searxng/issues/95
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/webapp.py:
```diff
@@ -686,6 +691,9 @@ def search():
     if output_format not in ['html', 'csv', 'json', 'rss']:
         output_format = 'html'
 
+    if output_format not in get_value(settings, 'search', 'formats', default=[]):
```
`default=[]` --> `default=['html', 'csv', 'json', 'rss']` otherwise it will break existing setup (without `search.formats`).

Looking at the code above:

```python
OUTPUT_FORMATS = ['html', 'csv', 'json', 'rss']

...
    allowed_formats = get_value(settings, 'search', 'formats', default=OUTPUT_FORMATS)
    output_format = request.form.get('format')
    if output_format not in OUTPUT_FORMATS:
        # undefined or unknown format: fallback to the first allowed format
        # default setting: 'html' 
        output_format  = allowed_formats[0]
    elif output_format not in allowed_formats:
        # the format is valid but not allowed: error 403
        flask.abort(403)
```

---

The fallback part keeps the API compatibility (`if output_format not in OUTPUT_FORMATS`)

return42 replied:
> otherwise it will break existing setup (without search.formats).

good idea to set the default including all formats .. but should we set the default in the code above or is it better placed nearby loading the settings?

Next thought: is it good to have defaults in settings.yml and other defauts in the code (e.g. nearby loading settings).

What I want is a single point of definitions .. ATM this single point is the settings,yml .. should we really add a new place where defaults could be located? (hope I was clear)

dalf replied:
> placed nearby loading the settings

I don't understand the location.

A simple solution for now: a the top of webapp.py (?)

---

With the searx PR 2501:
* the default value is in `searx/settings_defaults.py`
* `allowed_formats = get_value(settings, 'search', 'formats', default=OUTPUT_FORMATS)` --> `allowed_formats = settings['search']['formats']` (we are sure there is a value)

As discuss in this PR, a Settings class would be better, then the Preferences can use a Settings instance.

return42 replied:
> I don't understand the location.

I mean a file like ..

> With the searx PR 2501:
>    the default value is in searx/settings_defaults.py

I think a solution like PR2501 will help ... 

> As discuss in this PR, a Settings class would be better, then the Preferences can use a Settings instance.

PR2501 also implements static_path handling .. for me it is a bit hard to merge all this into tthe current searxng without to fail ... If you have time, do you like to merge this PR?  If so, I would vote to implement a class.

dalf commented on searx/utils.py:
```diff
@@ -500,6 +501,62 @@ def get_engine_from_settings(name):
+            ret_val = default
+        else:
+            ret_val = get_value(ret_val, *keys, default=default)
+    return ret_val
```
A version without recursive call:

```python
def get_value(dictionary, *keys, default=NOT_EXISTS):
    obj = dictionary
    for k in keys:
        if not isinstance(obj, Mapping):
            raise TypeError("expected mapping type, got %s" % type(obj))
        obj = obj.get(k, default)
        if obj is default:
            return obj
    return obj
```

return42 replied:
> A version without recursive call:

Thanks, good suggestion / implemented in 61c3cf46

return42 committed:
[mod] utils.get_value() - avoidance of a recursion

In a comment [1] dalf suggested to avoid a recursion of get_value()

[1] https://github.com/searxng/searxng/pull/99#discussion_r640833716

Suggested-by: Alexandre Flament <alex@al-f.net>
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/webapp.py:
```diff
@@ -683,9 +693,12 @@ def search():
         output_format = 'html'
 
+    if output_format not in get_value(settings, 'search', 'formats', default=OUTPUT_FORMATS):
+        flask.abort(403)
```
Looking at the documentation: https://flask.palletsprojects.com/en/2.0.x/api/#flask.abort

There is no need for an exception, just `return 'Access denied', 403` is enough.

return42 replied:
Hm, there is a difference .. 

AFAIK:  just a `return 'Access denied', 403` will be handled by the flask app (Flask generates a 403 response with the given string).   `flask.abort(403)` will raise an exception and the 403 page comes from the HTTP server.

If I was unclear, try both versions in a local instance (make run).

The question for me is: what do we want, should flask build the response or should the 403 be propagated over the http server? I decided to give the 403 back to the http server.

dalf replied:
> the 403 page comes from the HTTP server.

So either Werkzeug or uwsgi ?

* https://uwsgi-docs.readthedocs.io/en/latest/Changelog-1.9.21.html#error-pages

>  I decided to give the 403 back to the http server.

why ?

what about: 

https://github.com/searxng/searxng/blob/b48b4c93d586f52418295a15a1c93934f007c41f/searx/webapp.py#L1106-L1112

?

return42 replied:
> So either Werkzeug or uwsgi ?

AFAIK: Whatever (error page) is configured in the http server. Does not depend on the technology which is used to bind the Flask-App.

>  why?

It was just my gut feeling: I thought is a **Forbidden** (403)  which is better passed upwards.

The places you quote are  **Bad Request**"s (400) .. means: _"the request could not be understood by the server due to malformed syntax."_ which is IMO different to "**Forbidden**".

I do not have much experiences and I don't want to judge it .. if you prefer 403 handled by the Flask-App it is OK for me .. if so, the  _"why?"_ left on my side :)

return42 replied:
I don't know if it helps, but Flask's documentation says: [Redirects and Errors: ... to abort a request early with an error code, use the abort() function:](https://flask.palletsprojects.com/en/2.0.x/quickstart/#redirects-and-errors)

dalf replied:
> AFAIK: Whatever (error page) is configured in the http server. Does not depend on the technology which is used to bind the Flask-App.

I wish to reduce the number of configuration locations.

But it returns an HTTP 403 whatever the solution.
